### PR TITLE
Use CMake to build on Windows.

### PR DESCRIPTION
**Pull request by The-King-of-Toasters**
Now, we all have our opinions on build systems (CMake isn't my first choice either), but I chose it because its well supported on Windows/*NIX, and it makes life easier for future homebrew ports. The Vita and devkitPro scenes have custom CMake toolchains/scripts that makes their lives easier when doing ports, and my hope is that these get mainlined here.

I'm marking it as a draft for now since there's still problems to hunt down, but I'm able to build SDL2/Sokol on Linux, macOS and Windows 11. On my 5800X3D + 4080 rig in Windows 11, i'm getting ~5000 fps on the first map.

The instructions for Windows at the moment are:

```sh
# install vcpkg
vcpkg install sdl2 glew # add :x64-windows to the end to get amd64 versions
vcpkg integrate install
path\to\cmake.exe \
	-DCMAKE_C_COMPILER=path\to\clang-cl.exe
	-B build-dir \
	-S path\to\wipeout-rewrite \
	"-DCMAKE_TOOLCHAIN_FILE=path\to\vcpkg\scripts\buildsystems\vcpkg.cmake"
```

`vcpkg` makes building easy, bit manually specifying the paths to the sdl2/glew MSVC builds using `CMAKE_PREFIX_PATH` should also work. `clang-cl` is used because of the usage of C99 stuff (e.g. `__typeof__`) that MSVC still doesn't support, even in C11 mode. I also "borrowed" the `.ico` of the ATI Rage release and added it here.

For macOS, it's:

```sh
cmake \
	-B build-dir
	-S path/to/wipeout-rewrite
	-DCMAKE_PREFIX_PATH="$(brew --prefix sdl2)"
```

### CODE REVIEW

File: src/utils.h [hunk 1]
```diff
@@ -4,6 +4,12 @@
 #include <string.h>
 #include "types.h"
 
+#ifdef WIN32
+	#undef min
```
Comment (The-King-of-Toasters) [inline src/utils.h:8]: Despite defining `NOMINMAX` earlier, I'm still getting errors about `min` and `max` already being defined.

File: src/utils.h [hunk 2]
```diff
@@ -4,6 +4,12 @@
 #include <string.h>
 #include "types.h"
 
+#ifdef WIN32
```
Comment (The-King-of-Toasters) [inline src/utils.h:7]: Despite defining `NOMINMAX` earlier, I'm still getting errors about `min` and `max` already being defined.

File: src/render_gl.c [hunk 3]
```diff
@@ -2,7 +2,7 @@
 // macOS
 #if defined(__APPLE__) && defined(__MACH__)
 	#include <OpenGL/gl.h>
-	#include <OpenGL/glext.h>
+	#include <OpenGL/gl3.h>
```
Comment (The-King-of-Toasters) [inline src/render_gl.c:5]: The CRT shader additions required this header specifically. Also, is there a reason why macOS builds don't use `GLEW`?

**Comment by sajattack** [comment]
I would recommend creating a vcpkg.json file so packages are automatically picked up by visual studio, rather than instructing people to install packages manually.

File: src/utils.h (hunk 2 above)
Comment (The-King-of-Toasters) [reply src/utils.h:7]: I had a deeper look at this and found out where these errors are coming from. In the "Universal C Runtime" headers, a different set of `min` and `max` macros are defined in `stdlib.h`. The logic is basically:

```c
#if defined(_CRT_INTERNAL_NONSTDC_NAMES) && _CRT_INTERNAL_NONSTDC_NAMES
    #ifndef __cplusplus
        #define max(a,b) (((a) > (b)) ? (a) : (b))
        #define min(a,b) (((a) < (b)) ? (a) : (b))
    #endif
```

Setting `_CRT_INTERNAL_NONSTDC_NAMES` stops this from happening, but this has the effect of not declaring the `stat` bits used in utils.c. This too can be worked around by using `GetFileAttributesA`, but at this point I'd rather just stick to `undef min`.

The alternative is to rename `min/max` to `MIN/MAX` and sidestep this issue completely.

**Comment by The-King-of-Toasters** [comment]
@phoboslab So this PR has been cooking for a while now, just wondering if this is a valid improvement for you? Any comments are appreciated here.

**Comment by phoboslab** [comment]
Sorry for my lack of comments so far.

I'm not a big fan of build systems in general and would like to keep it as simple as possible. I realize the necessity for something more cross-platform than `make` and `cmake` is probably the next best choice – as you stated. So, yes, I think this all makes sense and is worth merging once it's ready.

That said, my knowledge of cmake is abysmal and I haven't used Windows in a while. Couple of questions; pardon my ignorance:

- why is specifying the build and source dir (`-B`, `-S`) necessary? Isn't the convention for cmake `mkdir build && cd build && cmake ..`?
- for me `cmake -B ./build/ -S ./` produces a makefile that compiles, but doesn't link. Somehow none of the required libraries end up in the Makefile (e.g. `-lm` is missing, even though I see that it is specified in the cmakelist.txt). What am I missing?
- how do I generate the Makefile for the WASM version with this?
- I believe `-Wpedantic` is a bit too much, considering that it currently produces a lot of warnings, drowning the "important" ones.

Going forward with this, we should probably remove the current `Makefile` and also replace the build instructions in the readme.

**Comment by The-King-of-Toasters** [comment]
`-B` and `-S` are just "modern" replacements for the classic idiom of:

```sh
mkdir build
cd build
cmake ..
```

where `build` and `..` are the build/source paths, respectively. The old way still works fine.

As for the Makefile build, I actually haven't tested that since I've been building with Ninja (`-G Ninja`). Since you mentioned not linking `libm` I presume you're building on Linux?

For WASM builds, you run emcmake as the first parameter and then the usual cmake command line afterwards:

```sh
emcmake cmake ...
```

Unfortunately I hit a snag with trying to replicate full/minimal builds. Basically, each build generates the same `wipeout.js` file, with the only difference being in the file package section. I'll add another comment afterwards with a better explaination. Right now you can pass `-DMINIMAL_BUNDLE` to CMake to do a minimal web build.

**Comment by The-King-of-Toasters** [comment]
BTW, these are the platforms I've successfully built using CMake:

- Windows
	- clang_cl x86_64 using vcpkg.
	- clang_cl x86 using vcpkg.
	- clang_cl aarch64 using vcpkg and the OpenGL Compatibility Pack.
	- clang_cl x86_64 by setting `CMAKE_PREFIX_PATH` to prebuilt SDL2/glew directories.
	- MSYS2 + GCC 12 x86_64 ucrt.
- macOS
	- AppleClang aarch64 using homebrew.
- Linux
	- GCC 13 on Arch Linux using `sdl2/glew` packages.
	- Mingw-w64 + GCC 13 on Arch Linux using Mingw `sdl2/glew` packages.

**Comment by The-King-of-Toasters** [comment]
@phoboslab I added a comprehensive build guide that replaces the one in the readme.

I also cannot reproduce your build error on Linux, can you specify the commands you used to setup CMake and build?

**Comment by phoboslab** [comment]
Very cool; good job with the Building.md, too! Not sure what I was doing wrong previously, but building with cmake works for me now.

Remarks:
- Some case inconsistencies for build options between the cmakelist and Building.md: `SOFTWARE` vs. `Software`, `ON` vs. `On`, `OFF` vs. `Off`. Since these are forwarded to the C-preprocessor, they should be all uppercase!?
- the name of the executable is `wipeout` which of course is fine on Windows as `wipeout.exe` but clashes on Linux with the `wipeout/` directory name. I suggest you change the executable to `wipegame` for now. I'll maybe change the directory name to `data/` at a later date...
- testing: what is the usual process here? Do I move the data directory (`wipeout/`) into the build dir? Again, pardon my ignorance: previously I could just `make && ./wipegame` for testing. If I rename the executable to `wipegame` and move the `wipeout/` into the build dir and `cmake --build ./build && ./build/wipegame` that still works for the SDL2 build, but not for Sokol, which then searches for the data in the wrong place. What's the usual process here?

For me, the WASM build and iterating the native builds is more ergonomic with the current Makefile. I don't particularly like the idea of maintaining two different build systems, but I don't want to drag this PR out unnecessarily either. So (at least for some time) we should probably keep the old Makefile around after all.

**Comment by The-King-of-Toasters** [comment]
Regarding all points:

- CMake options are different to C preprocessor definitions. The latter only get touched right now using `target_compile_definitions`.
- I changed the target exe name from `wipegame` to `wipeout` because I always did out-of-tree builds, and I would run the executable with the `cwd` set to the source directory. It seems #108cfac changed this so that assets are loaded from the base directory of the exe, which worked before since `wipegame` was written to the source directory. IMO the game saves and the assets should always be located in the prefpath and `PATH_ASSETS/PATH_DATA` to be used for dev builds.
- I mentioned that I was going to explain the WASM situation in a previous comment, which I haven't yet done. I'll do it in a new issue since this one is closed.

**Pull request merged**